// @testing-library/user-event/dist/esm/setup/setup.js
import { prepareDocument } from '../document/index.js';
import { bindDispatchUIEvent } from '../event/index.js';
import { defaultOptionsSetup, defaultOptionsDirect } from '../options.js';
import '../utils/click/isClickableInput.js';
import { attachClipboardStubToView } from '../utils/dataTransfer/Clipboard.js';
import '../utils/edit/maxLength.js';
import '../utils/edit/isEditable.js';
import '@testing-library/dom';
import '@testing-library/dom/dist/helpers.js';
import '../utils/keyDef/readNextDescriptor.js';
import { getDocumentFromNode } from '../utils/misc/getDocumentFromNode.js';
import { setLevelRef, ApiLevel } from '../utils/misc/level.js';
import { wait } from '../utils/misc/wait.js';
import { System } from '../system/index.js';
import { Config } from './config.js';
import * as api from './api.js';
import { wrapAsync } from './wrapAsync.js';

function createConfig(options = {}, defaults = defaultOptionsSetup, node) {
    const document = getDocument(options, node, defaults);
    return {
        ...defaults,
        ...options,
        document,
        system: options.system ?? new System()
    };
}
/**
 * Start a "session" with userEvent.
 * All APIs returned by this function share an input device state and a default configuration.
 */ function setupMain(options = {}) {
    const config = createConfig(options);
    prepareDocument(config.document);
    const view = config.document.defaultView ?? /* istanbul ignore next */ globalThis.window;
    attachClipboardStubToView(view);
    return doSetup(config);
}
/**
 * Setup in direct call per `userEvent.anyApi()`
 */ function setupDirect({ keyboardState , pointerState , ...options } = {}, node) {
    const config = createConfig({
        ...options,
        system: pointerState ?? keyboardState
    }, defaultOptionsDirect, node);
    prepareDocument(config.document);
    return {
        config,
        api: doSetup(config)
    };
}
/**
 * Create a set of callbacks with different default settings but the same state.
 */ function setupSub(options) {
    return doSetup({
        ...this[Config],
        ...options
    });
}
function wrapAndBindImpl(instance, impl) {
    function method(...args) {
        setLevelRef(instance[Config], ApiLevel.Call);
        return wrapAsync(()=>impl.apply(instance, args).then(async (ret)=>{
                await wait(instance[Config]);
                return ret;
            }));
    }
    Object.defineProperty(method, 'name', {
        get: ()=>impl.name
    });
    return method;
}
function doSetup(config) {
    const instance = {
        [Config]: config,
        dispatchUIEvent: bindDispatchUIEvent(config),
        ...api
    };
    return {
        ...Object.fromEntries(Object.entries(api).map(([name, api])=>[
                name,
                wrapAndBindImpl(instance, api), 
            ])),
        setup: setupSub.bind(instance)
    };
}
function getDocument(options, node, defaults) {
    return (options.document ?? (node && getDocumentFromNode(node))) ?? defaults.document;
}

export { createConfig, setupDirect, setupMain, setupSub };
