// @testing-library/user-event/dist/cjs/event/createEvent.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../utils/click/isClickableInput.js');
require('../utils/dataTransfer/Clipboard.js');
require('./behavior/click.js');
require('./behavior/cut.js');
require('./behavior/keydown.js');
require('./behavior/keypress.js');
require('./behavior/keyup.js');
require('./behavior/paste.js');
require('@testing-library/dom');
var eventMap = require('./eventMap.js');
require('../utils/edit/maxLength.js');
require('../utils/edit/isEditable.js');
var getWindow = require('../utils/misc/getWindow.js');
require('../utils/keyDef/readNextDescriptor.js');
require('../utils/misc/level.js');
require('../options.js');

const eventInitializer = {
    ClipboardEvent: [
        initClipboardEvent
    ],
    InputEvent: [
        initUIEvent,
        initInputEvent
    ],
    MouseEvent: [
        initUIEvent,
        initUIEventModififiers,
        initMouseEvent
    ],
    PointerEvent: [
        initUIEvent,
        initUIEventModififiers,
        initMouseEvent,
        initPointerEvent, 
    ],
    KeyboardEvent: [
        initUIEvent,
        initUIEventModififiers,
        initKeyboardEvent
    ]
};
function createEvent(type, target, init) {
    const window = getWindow.getWindow(target);
    const { EventType , defaultInit  } = eventMap.eventMap[eventMap.eventMapKeys[type]];
    const event = new (getEventConstructors(window))[EventType](type, defaultInit);
    eventInitializer[EventType]?.forEach((f)=>f(event, init ?? {}));
    return event;
}
/* istanbul ignore next */ function getEventConstructors(window) {
    /* eslint-disable @typescript-eslint/no-unnecessary-condition, @typescript-eslint/no-extraneous-class */ const Event = window.Event ?? class Event {
    };
    const AnimationEvent = window.AnimationEvent ?? class AnimationEvent extends Event {
    };
    const ClipboardEvent = window.ClipboardEvent ?? class ClipboardEvent extends Event {
    };
    const PopStateEvent = window.PopStateEvent ?? class PopStateEvent extends Event {
    };
    const ProgressEvent = window.ProgressEvent ?? class ProgressEvent extends Event {
    };
    const TransitionEvent = window.TransitionEvent ?? class TransitionEvent extends Event {
    };
    const UIEvent = window.UIEvent ?? class UIEvent extends Event {
    };
    const CompositionEvent = window.CompositionEvent ?? class CompositionEvent extends UIEvent {
    };
    const FocusEvent = window.FocusEvent ?? class FocusEvent extends UIEvent {
    };
    const InputEvent = window.InputEvent ?? class InputEvent extends UIEvent {
    };
    const KeyboardEvent = window.KeyboardEvent ?? class KeyboardEvent extends UIEvent {
    };
    const MouseEvent = window.MouseEvent ?? class MouseEvent extends UIEvent {
    };
    const DragEvent = window.DragEvent ?? class DragEvent extends MouseEvent {
    };
    const PointerEvent = window.PointerEvent ?? class PointerEvent extends MouseEvent {
    };
    const TouchEvent = window.TouchEvent ?? class TouchEvent extends UIEvent {
    };
    /* eslint-enable @typescript-eslint/no-unnecessary-condition, @typescript-eslint/no-extraneous-class */ return {
        Event,
        AnimationEvent,
        ClipboardEvent,
        PopStateEvent,
        ProgressEvent,
        TransitionEvent,
        UIEvent,
        CompositionEvent,
        FocusEvent,
        InputEvent,
        KeyboardEvent,
        MouseEvent,
        DragEvent,
        PointerEvent,
        TouchEvent
    };
}
function assignProps(obj, props) {
    for (const [key, value] of Object.entries(props)){
        Object.defineProperty(obj, key, {
            get: ()=>value ?? null
        });
    }
}
function sanitizeNumber(n) {
    return Number(n ?? 0);
}
function initClipboardEvent(event, { clipboardData  }) {
    assignProps(event, {
        clipboardData
    });
}
function initInputEvent(event, { data , inputType , isComposing  }) {
    assignProps(event, {
        data,
        isComposing: Boolean(isComposing),
        inputType: String(inputType)
    });
}
function initUIEvent(event, { view , detail  }) {
    assignProps(event, {
        view,
        detail: sanitizeNumber(detail ?? 0)
    });
}
function initUIEventModififiers(event, { altKey , ctrlKey , metaKey , shiftKey , modifierAltGraph , modifierCapsLock , modifierFn , modifierFnLock , modifierNumLock , modifierScrollLock , modifierSymbol , modifierSymbolLock  }) {
    assignProps(event, {
        altKey: Boolean(altKey),
        ctrlKey: Boolean(ctrlKey),
        metaKey: Boolean(metaKey),
        shiftKey: Boolean(shiftKey),
        getModifierState (k) {
            return Boolean({
                Alt: altKey,
                AltGraph: modifierAltGraph,
                CapsLock: modifierCapsLock,
                Control: ctrlKey,
                Fn: modifierFn,
                FnLock: modifierFnLock,
                Meta: metaKey,
                NumLock: modifierNumLock,
                ScrollLock: modifierScrollLock,
                Shift: shiftKey,
                Symbol: modifierSymbol,
                SymbolLock: modifierSymbolLock
            }[k]);
        }
    });
}
function initKeyboardEvent(event, { key , code , location , repeat , isComposing , charCode  }) {
    assignProps(event, {
        key: String(key),
        code: String(code),
        location: sanitizeNumber(location),
        repeat: Boolean(repeat),
        isComposing: Boolean(isComposing),
        charCode
    });
}
function initMouseEvent(event, { x , y , screenX , screenY , clientX =x , clientY =y , button , buttons , relatedTarget  }) {
    assignProps(event, {
        screenX: sanitizeNumber(screenX),
        screenY: sanitizeNumber(screenY),
        clientX: sanitizeNumber(clientX),
        x: sanitizeNumber(clientX),
        clientY: sanitizeNumber(clientY),
        y: sanitizeNumber(clientY),
        button: sanitizeNumber(button),
        buttons: sanitizeNumber(buttons),
        relatedTarget
    });
}
function initPointerEvent(event, { pointerId , width , height , pressure , tangentialPressure , tiltX , tiltY , twist , pointerType , isPrimary  }) {
    assignProps(event, {
        pointerId: sanitizeNumber(pointerId),
        width: sanitizeNumber(width),
        height: sanitizeNumber(height),
        pressure: sanitizeNumber(pressure),
        tangentialPressure: sanitizeNumber(tangentialPressure),
        tiltX: sanitizeNumber(tiltX),
        tiltY: sanitizeNumber(tiltY),
        twist: sanitizeNumber(twist),
        pointerType: String(pointerType),
        isPrimary: Boolean(isPrimary)
    });
}

exports.createEvent = createEvent;
